spinn_front_end_common/interface/provenance/fec_timer.py

```diff
@@ -172,23 +209,79 @@ def _stop_timer(self):
-        time_taken = self._stop_timer()
+        timedelta = self._stop_timer()
         if exc_type is None:
-            message = f"{self._algorithm} took {time_taken} "
+            message = f"{self._algorithm} took {timedelta} "
             skip = None
         else:
             try:
                 message = (f"{self._algorithm} exited with "
-                           f"{exc_type.__name__} after {time_taken}")
+                           f"{exc_type.__name__} after {timedelta}")
                 skip = exc_type.__name__
             except Exception as ex:  # pylint: disable=broad-except
                 message = f"{self._algorithm} exited with an exception" \
-                          f"after {time_taken}"
+                          f"after {timedelta}"
                 skip = f"Exception {ex}"
 
         with ProvenanceWriter() as db:
-            db.insert_timing(
-                self._category, self._algorithm, time_taken.microseconds,
-                _simulator.n_loops, skip)
+            db.insert_timing(self._category_id, self._algorithm, self._work,
+                             timedelta, skip)
         self._report(message)
         return False
+
+    @classmethod
+    def _change_category(cls, category_name):
+        """
+        This method should only be called via the View!
+
+        :param str category: Category to switch to
+        """
+        time_now = _now()
+        with ProvenanceWriter() as db:
+            if cls._category_id:
+                diff = _convert_to_timedelta(time_now - cls._category_time)
+                db.insert_category_timing(cls._category_id, diff)
+            cls._category_id = db.insert_category(
+                category_name, cls._machine_on)
+        cls._category_name = category_name
+        cls._category_time = time_now
+
+    @classmethod
+    def start_category(cls, category_name, machine_on=None):
+        """
+        This method should only be called via the View!
+
+        :param SimulatorStage category: Stage to switch to
+        :param machine_on: What to change machine on too.
+            Or None to leave as is
+        :type machine_on: None or bool
+        """
+        cls._previous.append(cls._category_name)
+        cls._change_category(category_name)
+        if machine_on is not None:
+            cls._machine_on = machine_on
+
+    @classmethod
+    def end_category(cls, category_name):
+        """
+        This method should only be
+        called via the View!
+
+        :param SimulatorStage category: Stage to end
+        """
+        if cls._category_name != category_name:
+            if cls._category_name != cls.ERROR:
+                logger.exception(f"Incorrect end of category: {category_name}"
+                                 f" when in category{cls._category_name}")
+            cls._previous = []
+            cls._change_category(cls.ERROR)
+        else:
+            try:
+                previous = cls._previous.pop()
```

Comment:
Doesn't this mean that you need to have at least one category which you never end?  If so, would it make sense for that to go in the constructor as an option, or else in the _previous list by default?  It could then error if you try to take that out at any time.

Reply:
There should be a WAITING started during sim.setup.
(SHOULD! oops)
Before setup time is of zero interest.

It does error if you get out the last category but TOO silently.
Should at the very least log.exception

Reply:
There now is one category WAITING which starts during setup which is end_category is called will raise an Exception.

There is now also a stop_category_timing which cleanly closes down the running Category and returns the FecTimer into the state it was before setup was called. 
That works no matter how many start_category calls are waiting for  end_category